### PR TITLE
[BEAM-14326] Make sure BigQuery daemon thread doesn't exit suddenly, as this leads to pipeline stuckness

### PR DESCRIPTION
**Pull request by reuvenlax**

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableSchemaCache.java [hunk 1]
```diff
@@ -271,10 +280,11 @@ public void refreshThread() {
       if (timeRemaining.getMillis() > 0) {
         Thread.sleep(timeRemaining.getMillis());
       }
-    } catch (InterruptedException e) {
-      runUnderMonitor(() -> this.stopped = true);
-      return;
-    } catch (IOException e) {
+    } catch (Exception e) {
```
Comment (TheNeuralBit) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableSchemaCache.java:283]: Is it appropriate to swallow _all_ exceptions? Could there be cases where the daemon needs to be monitored and restarted?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsTableRow.java [hunk 2]
```diff
@@ -95,7 +96,14 @@ public MessageConverter<T> getMessageConverter(
                       + "using a create disposition of CREATE_IF_NEEDED.");
             }
           }
+        } else {
+          // Make sure we register this schema with the cache, unless there's already a more
+          // up-to-date schema.
+          tableSchema =
+              MoreObjects.firstNonNull(
+                  SCHEMA_CACHE.putSchemaIfAbsent(tableReference, tableSchema), tableSchema);
```
[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiDynamicDestinationsTableRow.java:104]: Yes - mostly applies to unit tests (where we explicitly clear the static caches in between tests) - I don't think this case would occur in in production.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableSchemaCache.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/TableSchemaCache.java:285]: That is sorta what is happening - we still exit the function, but schedule another closure. I'm not aware of any case where we want to shut down the executor itself.

**Review by kennknowles — COMMENTED** [review]
Between this as the other issues around this, it seems like we need some continuously-running integration tests.

**Comment by reuvenlax** [comment]
@kennknowles Yes we absolutely do, though note this issue was caught in unit test flakiness, and would be much less common in a production integration test.

**Comment by reuvenlax** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by reuvenlax** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by reuvenlax** [comment]
Any more comments?

**Review by TheNeuralBit — APPROVED** [review]
Looks like there were some (likely unrelated) issues in GitHub Actions jobs that I just restarted. Otherwise LGTM

**Pull request merged**